tests/e2e/http_health_check_test.go

```diff
@@ -0,0 +1,236 @@
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	require.NoErrorf(t, err, "failed to creat request %+v", err)
+	resp, herr := client.Do(req)
+	cancel()
+	if expectTimeoutError {
+		if herr != nil && strings.Contains(herr.Error(), context.DeadlineExceeded.Error()) {
+			return
+		}
+	}
+	require.NoErrorf(t, herr, "failed to get response %+v", err)
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	resp.Body.Close()
+	require.NoErrorf(t, err, "failed to read response %+v", err)
+
+	t.Logf("health check response body is: %s", body)
+	require.Equal(t, expectStatusCode, resp.StatusCode)
+}
+
+func triggerNoSpaceAlarm(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	buf := strings.Repeat("b", os.Getpagesize())
+	etcdctl := clus.Etcdctl()
+	for {
+		if err := etcdctl.Put(ctx, "foo", buf, config.PutOptions{}); err != nil {
+			if !strings.Contains(err.Error(), "etcdserver: mvcc: database space exceeded") {
+				t.Fatal(err)
+			}
+			break
+		}
+	}
+}
+
+func triggerSlowApply(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	// the following proposal will be blocked at applying stage
+	// because when apply index < committed index, linearizable read would time out.
+	checkFailPointEnabled(t, clus.Procs[0])
+	require.NoError(t, clus.Procs[0].Failpoints().SetupHTTP(ctx, "beforeApplyOneEntryNormal", `sleep("3s")`))
+	require.NoError(t, clus.Procs[1].Etcdctl().Put(ctx, "foo", "bar", config.PutOptions{}))
+}
+
+func blackhole(_ context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	member := clus.Procs[0]
+	proxy := member.PeerProxy()
+	t.Logf("Blackholing traffic from and to member %q", member.Config().Name)
+	proxy.BlackholeTx()
+	proxy.BlackholeRx()
+}
+
+func triggerRaftLoopDeadLock(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	checkFailPointEnabled(t, clus.Procs[0])
+	require.NoError(t, clus.Procs[0].Failpoints().SetupHTTP(ctx, "raftBeforeSave", `sleep("3s")`))
+	clus.Procs[0].Etcdctl().Put(context.Background(), "foo", "bar", config.PutOptions{})
+}
+
+func triggerSlowBufferWriteBackWithAuth(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	checkFailPointEnabled(t, clus.Procs[0])
+
+	etcdctl := clus.Etcdctl()
+	_, err := etcdctl.UserAdd(ctx, "root", "root", config.UserAddOptions{})
+	require.NoError(t, err)
+	_, err = etcdctl.UserGrantRole(ctx, "root", "root")
+	require.NoError(t, err)
+	require.NoError(t, etcdctl.AuthEnable(ctx))
+
+	require.NoError(t, clus.Procs[0].Failpoints().SetupHTTP(ctx, "beforeWritebackBuf", `sleep("3s")`))
+	clus.Procs[0].Etcdctl(e2e.WithAuth("root", "root")).Put(context.Background(), "foo", "bar", config.PutOptions{Timeout: 200 * time.Millisecond})
+}
+
+func checkFailPointEnabled(t *testing.T, member e2e.EtcdProcess) {
+	if !member.Failpoints().Enabled() {
+		t.Skip("please run `make gofail-enable` and compile etcd binary")
+	}
```

Comment:
Can we perform the check automatically in e2e.BeforeTest? https://github.com/etcd-io/etcd/blob/262a3a1208a74e7dc0719c2c75096e934242546b/tests/framework/e2e/testing.go#L23

Similar to integration.BeforeTest,
https://github.com/etcd-io/etcd/blob/262a3a1208a74e7dc0719c2c75096e934242546b/tests/framework/integration/testing.go#L100-L108

Reply:
We need to spin up gofail http server by `GOFAIL_HTTP="127.0.0.1:1234" ./etcd` first and send http request to determine if gofail is enabled. It sounds like over engineering to implement it in `e2e.BeforeTest`. 

I prefer keep the PR as it is unless you have better ideas. WDYT? @ahrtr

Reply:
At least let's move this function into e2e framework as a common function?

The logic should be something like below,

```
if cfg.GoFailEnabled && !member.Failpoints().Enabled() {
    t.Skip("xxxxx")
}
```

Reply:
Note we have an open issue and stale pr for robustness that relates to this:
- https://github.com/etcd-io/etcd/issues/15258
- https://github.com/etcd-io/etcd/pull/15963

Would be great if we could close out that issue as part of this pr if we are already looking at this area.

Reply:
> At least let's move this function into e2e framework as a common function?
> 
> The logic should be something like below,
> 
> ```
> if cfg.GoFailEnabled && !member.Failpoints().Enabled() {
>     t.Skip("xxxxx")
> }
> ```

Implemented.